File: ios/App/App/Billing/EverittBillingPlugin.swift
import Foundation
import Capacitor
import StoreKit
import UIKit

@objc(EverittBillingPlugin)
public class EverittBillingPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "EverittBilling"
    public let jsName = "EverittBilling"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "loadProducts", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "purchase", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "restorePurchases", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "manageSubscriptions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getProductIds", returnType: CAPPluginReturnPromise)
    ]

    private let defaultProductIds = [
        "com.everittventures.everittos.pro.monthly.v3",
        "com.everittventures.everittos.business.monthly.v2",
        "com.everittventures.everittos.starter.monthly.v4",
        "com.everittventures.everittos.growth.monthly.v2",
        "com.everittventures.everittos.enterprise.monthly.v2"
    ]

    @objc func loadProducts(_ call: CAPPluginCall) {
        let ids = call.getArray("productIds", String.self) ?? defaultProductIds
        Task {
            do {
                let products = try await StoreKitBillingManager.shared.loadProducts(productIDs: Set(ids))
                let mapped = products.map { product -> [String: Any] in
                    [
                        "productId": product.id,
                        "title": product.displayName,
                        "description": product.description,
                        "price": product.displayPrice,
                        "priceCurrencyCode": product.priceFormatStyle.currencyCode,
                        "billingPeriod": "monthly"
                    ]
                }
                call.resolve(["products": mapped])
            } catch {
                call.reject("Store products are temporarily unavailable.")
            }
        }
    }

    @objc func purchase(_ call: CAPPluginCall) {
        guard let productId = call.getString("productId"), !productId.isEmpty else {
            call.reject("This plan is temporarily unavailable.")
            return
        }

        Task {
            do {
                var products = await StoreKitBillingManager.shared.products
                if !products.contains(where: { $0.id == productId }) {
                    products = try await StoreKitBillingManager.shared.loadProducts(productIDs: [productId])
                }
                guard let product = products.first(where: { $0.id == productId }) else {
                    call.reject("This plan is temporarily unavailable.")
                    return
                }
                let payload = try await StoreKitBillingManager.shared.purchase(product)
                call.resolve([
                    "productId": payload.productId,
                    "transactionId": payload.transactionId,
                    "originalTransactionId": payload.originalTransactionId,
                    "signedTransaction": payload.signedTransaction,
                    "pending": payload.pending,
                    "cancelled": payload.cancelled
                ])
            } catch {
                call.reject("We could not complete the purchase. Please try again.")
            }
        }
    }

    @objc func restorePurchases(_ call: CAPPluginCall) {
        Task {
            do {
                let purchases = try await StoreKitBillingManager.shared.restore()
                let mapped = purchases.map { purchase -> [String: Any] in
                    [
                        "productId": purchase.productId,
                        "transactionId": purchase.transactionId,
                        "originalTransactionId": purchase.originalTransactionId,
                        "signedTransaction": purchase.signedTransaction,
                        "pending": purchase.pending,
                        "cancelled": purchase.cancelled
                    ]
                }
                call.resolve(["purchases": mapped])
            } catch {
                call.reject("We could not restore purchases. Please try again.")
            }
        }
    }

    @objc func manageSubscriptions(_ call: CAPPluginCall) {
        Task {
            await StoreKitBillingManager.shared.manageSubscriptions()
            call.resolve(["opened": true])
        }
    }

    @objc func getProductIds(_ call: CAPPluginCall) {
        call.resolve(["productIds": defaultProductIds])
    }
}

File: android/app/src/main/java/com/everittventures/everittos/billing/EverittBillingPlugin.java
package com.everittventures.everittos.billing;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.net.Uri;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

@CapacitorPlugin(name = "EverittBilling")
public class EverittBillingPlugin extends Plugin {
    private PlayBillingManager billingManager;
    private PluginCall pendingPurchaseCall;
    private Consumer<PlayBillingModels.PurchasePayload> purchaseListener;

    private static final String[] DEFAULT_PRODUCT_IDS = new String[] {
            "everittos_pro",
            "everittos_business"
    };

    @Override
    public void load() {
        billingManager = new PlayBillingManager(getContext());
        purchaseListener = payload -> {
            PluginCall call = pendingPurchaseCall;
            pendingPurchaseCall = null;
            if (call == null) return;
            JSObject result = new JSObject();
            result.put("productId", payload.productId);
            result.put("purchaseToken", payload.purchaseToken);
            result.put("pending", payload.pending);
            result.put("cancelled", payload.cancelled);
            call.resolve(result);
        };
        billingManager.addPurchaseListener(purchaseListener);
    }

    @Override
    protected void handleOnDestroy() {
        if (billingManager != null && purchaseListener != null) {
            billingManager.removePurchaseListener(purchaseListener);
        }
        if (pendingPurchaseCall != null) {
            pendingPurchaseCall.reject("Purchase interrupted because the app closed.");
            pendingPurchaseCall = null;
        }
        super.handleOnDestroy();
    }

    @PluginMethod
    public void loadProducts(PluginCall call) {
        List<String> productIds = readProductIds(call);
        billingManager.queryProducts(productIds, products -> {
            JSArray array = new JSArray();
            for (Map<String, Object> product : products) {
                JSObject row = new JSObject();
                row.put("productId", String.valueOf(product.get("productId")));
                row.put("title", String.valueOf(product.get("title")));
                row.put("description", String.valueOf(product.get("description")));
                row.put("price", String.valueOf(product.get("price")));
                row.put("billingPeriod", String.valueOf(product.get("billingPeriod")));
                array.put(row);
            }
            JSObject result = new JSObject();
            result.put("products", array);
            call.resolve(result);
        }, call::reject);
    }

    @PluginMethod
    public void purchase(PluginCall call) {
        String productId = call.getString("productId", "");
        if (productId == null || productId.isEmpty()) {
            call.reject("productId is required");
            return;
        }
        if (pendingPurchaseCall != null) {
            call.reject("Another purchase is already in progress.");
            return;
        }
        Activity activity = getActivity();
        if (activity == null) {
            call.reject("Activity unavailable");
            return;
        }
        pendingPurchaseCall = call;
        billingManager.launchPurchase(activity, productId, error -> {
            PluginCall currentCall = pendingPurchaseCall;
            pendingPurchaseCall = null;
            if (currentCall != null) currentCall.reject(error);
        });
    }

    @PluginMethod
    public void restorePurchases(PluginCall call) {
        billingManager.queryExistingPurchases(purchases -> {
            JSArray array = new JSArray();
            for (PlayBillingModels.PurchasePayload purchase : purchases) {
                JSObject row = new JSObject();
                row.put("productId", purchase.productId);
                row.put("purchaseToken", purchase.purchaseToken);
                row.put("pending", purchase.pending);
                row.put("cancelled", purchase.cancelled);
                array.put(row);
            }
            JSObject result = new JSObject();
            result.put("purchases", array);
            call.resolve(result);
        }, call::reject);
    }

    @PluginMethod
    public void manageSubscriptions(PluginCall call) {
        String packageName = getContext().getPackageName();
        String url = "https://play.google.com/store/account/subscriptions?package=" + packageName;
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        try {
            getContext().startActivity(intent);
            JSObject result = new JSObject();
            result.put("opened", true);
            call.resolve(result);
        } catch (ActivityNotFoundException error) {
            call.reject("Google Play subscription management is unavailable on this device.");
        }
    }

    @PluginMethod
    public void getProductIds(PluginCall call) {
        JSArray array = new JSArray();
        for (String id : DEFAULT_PRODUCT_IDS) {
            array.put(id);
        }
        JSObject result = new JSObject();
        result.put("productIds", array);
        call.resolve(result);
    }

    private List<String> readProductIds(PluginCall call) {
        List<String> ids = new ArrayList<>();
        JSArray array = call.getArray("productIds");
        if (array != null) {
            for (int i = 0; i < array.length(); i++) {
                try {
                    ids.add(array.getString(i));
                } catch (Exception ignored) {
                    // skip
                }
            }
        }
        if (ids.isEmpty()) {
            for (String id : DEFAULT_PRODUCT_IDS) {
                ids.add(id);
            }
        }
        return ids;
    }
}
